src/core/numerics/pusher/boris.hpp

```diff
@@ -71,240 +70,192 @@ namespace core
-                // not strictly speaking this function's business
-                // to take advantage that we're already looping
-                // over rangeIn particles.
-                if (step == PushStep::PrePush)
-                {
-                    currentOut->charge = currentIn.charge;
-                    currentOut->weight = currentIn.weight;
-                    currentOut->v      = currentIn.v;
-                }
-                // push the particle
-                advancePosition_(currentIn, *currentOut);
-                ++currentOut;
+                outParticles[outIdx].charge = inParticles[inIdx].charge;
+                outParticles[outIdx].weight = inParticles[inIdx].weight;
+                outParticles[outIdx].v      = inParticles[inIdx].v;
             }
+            auto newCell = advancePosition_(inParticles[inIdx], outParticles[outIdx]);
+            if (newCell != inParticles[inIdx].iCell)
+                outParticles.change_icell(newCell, outIdx);
         }
+    }
 
-        template<typename ParticleRangeIn, typename ParticleRangeOut>
-        auto pushStep_(ParticleRangeIn const& rangeIn, ParticleRangeOut& rangeOut,
-                       ParticleSelector const& particleIsNotLeaving, PushStep step)
-        {
-            pushStep_(rangeIn, rangeOut, step);
-
-            // now all particles have been pushed
-            // those not satisfying the predicate after the push
-            // are found in [newEnd:end[
-            // those for which pred is true are in [firstOut,newEnd[
-            return std::partition(std::begin(rangeOut), std::end(rangeOut), particleIsNotLeaving);
-        }
 
 
+    /** Accelerate the particles in rangeIn and put the new velocity in rangeOut
+     */
+    void accelerate_(ParticleRange rangeIn, ParticleRange rangeOut, double mass)
+    {
+        double dto2m = 0.5 * dt_ / mass;
 
+        auto& inParticles  = rangeIn.array();
+        auto& outParticles = rangeOut.array();
 
-        /** Accelerate the particles in rangeIn and put the new velocity in rangeOut
-         */
-        template<typename ParticleRangeIn, typename ParticleRangeOut>
-        void accelerate_(ParticleRangeIn inputParticles, ParticleRangeOut outputParticles,
-                         double mass)
+        for (auto inIdx = rangeIn.ibegin(), outIdx = rangeOut.ibegin(); inIdx < rangeIn.iend();
+             ++inIdx, ++outIdx)
         {
-            double dto2m = 0.5 * dt_ / mass;
+            auto& inPart  = inParticles[inIdx];
+            auto& outPart = outParticles[inIdx];
+            double coef1  = inPart.charge * dto2m;
 
-            auto currentOut = outputParticles.begin();
+            // We now apply the 3 steps of the BORIS PUSHER
 
-            for (auto const& currentIn : inputParticles)
-            {
-                double coef1 = currentIn.charge * dto2m;
+            // 1st half push of the electric field
+            double velx1 = inPart.v[0] + coef1 * inPart.Ex;
```

Comment:
is it guaranteed that consequtive stack allocations are contiguous?
I'd feel a bit better if it was an array, but maybe it's the same

Reply:
you mean Ex, Ey, Ez ==> E[0..2] ?

Reply:
you're probably right, even though not really part of this PR (the diff is probably related to something else shifting lines)